arch/riscv/core/pmp.c

```diff
@@ -317,3 +322,254 @@ void z_riscv_pmp_stackguard_enable(struct k_thread *thread)
+			 remaining_partitions + 1, domain);
+	}
+	thread->arch.pmp_u_end_index = index;
+	thread->arch.pmp_u_update_nr = domain->arch.pmp_update_nr;
+
+	k_spin_unlock(&z_mem_domain_lock, key);
+}
+
+/**
+ * @brief Write PMP usermode content to actual PMP registers
+ *
+ * This is called on every context switch.
+ */
+void z_riscv_pmp_usermode_enable(struct k_thread *thread)
+{
+	struct k_mem_domain *domain = thread->mem_domain_info.mem_domain;
+
+	if (thread->arch.pmp_u_end_index == 0) {
+		/* z_riscv_pmp_usermode_prepare() has not been called yet */
+		return;
+	}
+
+	if (thread->arch.pmp_u_update_nr != domain->arch.pmp_update_nr) {
+		/*
+		 * Resynchronize our PMP entries with
+		 * the latest domain partition information.
+		 */
+		resync_pmp_domain(thread, domain);
+	}
+
+#ifdef CONFIG_PMP_STACK_GUARD
+	/* Make sure m-mode PMP usage is disabled before we reprogram it */
+	csr_clear(mstatus, MSTATUS_MPRV);
+#endif
+
+	/* Write our u-mode MPP entries */
+	write_pmp_entries(global_pmp_end_index, thread->arch.pmp_u_end_index,
+			  thread->arch.pmp_u_addr, thread->arch.pmp_u_cfg,
+			  true /* must clear to the end */);
+
+	if (PMP_DEBUG_DUMP) {
+		dump_pmp_regs("u-mode register dump");
+	}
+}
+
+int arch_mem_domain_max_partitions_get(void)
+{
+	int available_pmp_slots = CONFIG_PMP_SLOT;
+
+	/* remove those slots dedicated to global entries */
+	available_pmp_slots -= global_pmp_end_index;
+
+#if !defined(CONFIG_SMP)
+	/* One slot needed to map the is_user_mode variable */
+	available_pmp_slots -= 1;
+#endif
+
+	/* At least one slot to map the user thread's stack */
+	available_pmp_slots -= 1;
+
+	/*
+	 * Each partition may require either 1 or 2 PMP slots depending
+	 * on a couple factors that are not known in advance. Even when
+	 * arch_mem_domain_partition_add() is called, we can't tell if a
+	 * given partition will fit in the remaining PMP slots of an
+	 * affected thread if it hasn't executed in usermode yet.
+	 *
+	 * Give the most optimistic answer here (which should be pretty
+	 * accurate if MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT=y) and be
+	 * prepared to deny availability in resync_pmp_domain() if this
+	 * estimate was too high.
+	 */
+	return available_pmp_slots;
+}
+
+int arch_mem_domain_init(struct k_mem_domain *domain)
+{
+	domain->arch.pmp_update_nr = 0;
+	return 0;
+}
+
+int arch_mem_domain_partition_add(struct k_mem_domain *domain,
+				  uint32_t partition_id)
+{
+	/* Force resynchronization for every thread using this domain */
+	domain->arch.pmp_update_nr += 1;
+	return 0;
+}
+
+int arch_mem_domain_partition_remove(struct k_mem_domain *domain,
+				     uint32_t partition_id)
+{
+	/* Force resynchronization for every thread using this domain */
+	domain->arch.pmp_update_nr += 1;
+	return 0;
+}
+
+int arch_mem_domain_thread_add(struct k_thread *thread)
+{
+	/* Force resynchronization for this thread */
+	thread->arch.pmp_u_update_nr = 0;
+	return 0;
+}
+
+int arch_mem_domain_thread_remove(struct k_thread *thread)
+{
+	return 0;
```

Comment:
Why can this be empty?

Reply:
Because there is nothing to do here really. If the same thread is re-added to the same domain then its sequence number will be reset to zero and won't match the domain's sequence number which will force a resync the next time it is cheduled.